<a href="https://colab.research.google.com/github/arthurt1/Avalia-o-de-Modelos-de-ML/blob/rodrigo/Avalia%C3%A7%C3%A3o_de_Modelos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Nomes da Equipe:



*   Arthur Thomé Costa - 565756
*   Rodrigo Sousa Barbosa - 566581
*   Ryan dos Santos Oliveira - 565120




# Objetivos:
*   Aplicar diferentes métricas de avaliação para classificação, regressão e clusterização
*  Comparar o desempenho de múltiplos algoritmos
* Interpretar corretamente os resultados das métricas
* Identificar problemas como overfitting e underfitting

# Pequeno resumo do trabalho e resultados:

Preparação do Ambiente Virtual:

In [9]:
## Importação das bibliotecas
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from collections import Counter
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.datasets import load_iris, load_wine, fetch_openml
from sklearn.cluster import KMeans

# Parte 1: CLASSIFICAÇÃO


---


Dataset: *Iris Dataset*

Descrição: Classificação de espécies de flores Iris (Setosa, Versicolor, Virginica) baseado em medidas das sépalas e pétalas.

## Tarefa 1.1: Implementação e Avaliação Básica

Implemente o classificador K-Nearest Neighbors (KNN) com k=3 e k=5

### a) Divida o dataset em 70% treino e 30% teste. Para cada modelo, calcule:

- Acurácia
- Precisão (macro e micro)
- Recall (macro e micro)
- F1-score (macro e micro)
- Matriz de confusão

In [10]:
iris = load_iris()

df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['target'] = iris.target

df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


In [11]:
from sklearn.neighbors import KNeighborsClassifier


# Define features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# Split the dataset into 70% training and 30% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Scale the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("--- K-Nearest Neighbors (k=3) ---")

# Implement KNN with k=3
knn_3 = KNeighborsClassifier(n_neighbors=3)
knn_3.fit(X_train_scaled, y_train)
y_pred_knn_3 = knn_3.predict(X_test_scaled)

# Calculate metrics for k=3
accuracy_knn_3 = accuracy_score(y_test, y_pred_knn_3)
precision_macro_knn_3 = precision_score(y_test, y_pred_knn_3, average='macro')
precision_micro_knn_3 = precision_score(y_test, y_pred_knn_3, average='micro')
recall_macro_knn_3 = recall_score(y_test, y_pred_knn_3, average='macro')
recall_micro_knn_3 = recall_score(y_test, y_pred_knn_3, average='micro')
f1_macro_knn_3 = f1_score(y_test, y_pred_knn_3, average='macro')
f1_micro_knn_3 = f1_score(y_test, y_pred_knn_3, average='micro')
confusion_matrix_knn_3 = confusion_matrix(y_test, y_pred_knn_3)

print(f"Accuracy (k=3): {accuracy_knn_3:.4f}")
print(f"Precision (Macro, k=3): {precision_macro_knn_3:.4f}")
print(f"Precision (Micro, k=3): {precision_micro_knn_3:.4f}")
print(f"Recall (Macro, k=3): {recall_macro_knn_3:.4f}")
print(f"Recall (Micro, k=3): {recall_micro_knn_3:.4f}")
print(f"F1-score (Macro, k=3): {f1_macro_knn_3:.4f}")
print(f"F1-score (Micro, k=3): {f1_micro_knn_3:.4f}")
print("Confusion Matrix (k=3):")
print(confusion_matrix_knn_3)

print("\n--- K-Nearest Neighbors (k=5) ---")

# Implement KNN with k=5
knn_5 = KNeighborsClassifier(n_neighbors=5)
knn_5.fit(X_train_scaled, y_train)
y_pred_knn_5 = knn_5.predict(X_test_scaled)

# Calculate metrics for k=5
accuracy_knn_5 = accuracy_score(y_test, y_pred_knn_5)
precision_macro_knn_5 = precision_score(y_test, y_pred_knn_5, average='macro')
precision_micro_knn_5 = precision_score(y_test, y_pred_knn_5, average='micro')
recall_macro_knn_5 = recall_score(y_test, y_pred_knn_5, average='macro')
recall_micro_knn_5 = recall_score(y_test, y_pred_knn_5, average='micro')
f1_macro_knn_5 = f1_score(y_test, y_pred_knn_5, average='macro')
f1_micro_knn_5 = f1_score(y_test, y_pred_knn_5, average='micro')
confusion_matrix_knn_5 = confusion_matrix(y_test, y_pred_knn_5)

print(f"Accuracy (k=5): {accuracy_knn_5:.4f}")
print(f"Precision (Macro, k=5): {precision_macro_knn_5:.4f}")
print(f"Precision (Micro, k=5): {precision_micro_knn_5:.4f}")
print(f"Recall (Macro, k=5): {recall_macro_knn_5:.4f}")
print(f"Recall (Micro, k=5): {recall_micro_knn_5:.4f}")
print(f"F1-score (Macro, k=5): {f1_macro_knn_5:.4f}")
print(f"F1-score (Micro, k=5): {f1_micro_knn_5:.4f}")
print("Confusion Matrix (k=5):")
print(confusion_matrix_knn_5)


--- K-Nearest Neighbors (k=3) ---
Accuracy (k=3): 1.0000
Precision (Macro, k=3): 1.0000
Precision (Micro, k=3): 1.0000
Recall (Macro, k=3): 1.0000
Recall (Micro, k=3): 1.0000
F1-score (Macro, k=3): 1.0000
F1-score (Micro, k=3): 1.0000
Confusion Matrix (k=3):
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]

--- K-Nearest Neighbors (k=5) ---
Accuracy (k=5): 1.0000
Precision (Macro, k=5): 1.0000
Precision (Micro, k=5): 1.0000
Recall (Macro, k=5): 1.0000
Recall (Micro, k=5): 1.0000
F1-score (Macro, k=5): 1.0000
F1-score (Micro, k=5): 1.0000
Confusion Matrix (k=5):
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]


### b) Qual modelo obteve melhor desempenho geral? Justifique sua resposta considerando múltiplas métricas.

Ambos os modelos K-Nearest Neighbors, com k=3 e k=5, obtiveram desempenho idêntico e perfeito em todas as métricas de avaliação (Acurácia, Precisão, Recall e F1-score, tanto macro quanto micro), além de apresentarem a mesma Matriz de Confusão. Isso significa que, para este conjunto de dados de teste específico, ambos os modelos foram capazes de classificar todas as instâncias corretamente, resultando em 100% de acerto. Portanto, não há um desempenho superior de um modelo sobre o outro neste cenário.

### c) Existe alguma classe que é mais difícil de classificar? Como você identificou isso?

Analisando as Matrizes de Confusão para ambos os modelos (k=3 e k=5):

```
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]
```

Podemos observar que todos os valores fora da diagonal principal são zero. Isso indica que não houve nenhuma classificação incorreta para nenhuma das classes (0, 1 ou 2). Todos os 19 exemplos da classe 0, os 13 exemplos da classe 1 e os 13 exemplos da classe 2 foram corretamente identificados.

Portanto, para este conjunto de dados de teste, **não existe nenhuma classe que tenha sido mais difícil de classificar**, pois ambos os modelos alcançaram uma precisão perfeita para todas as espécies de Iris.

## Tarefa 1.2: Validação Cruzada

### a)Implemente validação cruzada k-fold com k=5 para todos os modelos da Tarefa 1.1.

In [12]:
from sklearn.model_selection import StratifiedKFold

# Initialize lists to store metrics for k=3 KNN
accuracy_scores_knn_3_cv = []
precision_macro_scores_knn_3_cv = []
precision_micro_scores_knn_3_cv = []
recall_macro_scores_knn_3_cv = []
recall_micro_scores_knn_3_cv = []
f1_macro_scores_knn_3_cv = []
f1_micro_scores_knn_3_cv = []

# Initialize lists to store metrics for k=5 KNN
accuracy_scores_knn_5_cv = []
precision_macro_scores_knn_5_cv = []
precision_micro_scores_knn_5_cv = []
recall_macro_scores_knn_5_cv = []
recall_micro_scores_knn_5_cv = []
f1_macro_scores_knn_5_cv = []
f1_micro_scores_knn_5_cv = []

# Instantiate StratifiedKFold
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

print("--- K-Nearest Neighbors (k=3) with 5-Fold Cross-Validation ---")

# Loop through each fold for k=3
for fold, (train_index, val_index) in enumerate(skf.split(X_train_scaled, y_train)):
    X_sub_train, X_val = X_train_scaled[train_index], X_train_scaled[val_index]
    y_sub_train, y_val = y_train.iloc[train_index], y_train.iloc[val_index]

    # Initialize and train KNN (k=3)
    knn_3_cv = KNeighborsClassifier(n_neighbors=3)
    knn_3_cv.fit(X_sub_train, y_sub_train)
    y_pred_knn_3_cv = knn_3_cv.predict(X_val)

    # Calculate metrics for the current fold
    accuracy = accuracy_score(y_val, y_pred_knn_3_cv)
    precision_macro = precision_score(y_val, y_pred_knn_3_cv, average='macro', zero_division=0)
    precision_micro = precision_score(y_val, y_pred_knn_3_cv, average='micro', zero_division=0)
    recall_macro = recall_score(y_val, y_pred_knn_3_cv, average='macro', zero_division=0)
    recall_micro = recall_score(y_val, y_pred_knn_3_cv, average='micro', zero_division=0)
    f1_macro = f1_score(y_val, y_pred_knn_3_cv, average='macro', zero_division=0)
    f1_micro = f1_score(y_val, y_pred_knn_3_cv, average='micro', zero_division=0)

    # Store metrics
    accuracy_scores_knn_3_cv.append(accuracy)
    precision_macro_scores_knn_3_cv.append(precision_macro)
    precision_micro_scores_knn_3_cv.append(precision_micro)
    recall_macro_scores_knn_3_cv.append(recall_macro)
    recall_micro_scores_knn_3_cv.append(recall_micro)
    f1_macro_scores_knn_3_cv.append(f1_macro)
    f1_micro_scores_knn_3_cv.append(f1_micro)

    print(f"\nFold {fold+1}:")
    print(f"  Accuracy: {accuracy:.4f}")
    print(f"  Precision (Macro): {precision_macro:.4f}")
    print(f"  Precision (Micro): {precision_micro:.4f}")
    print(f"  Recall (Macro): {recall_macro:.4f}")
    print(f"  Recall (Micro): {recall_micro:.4f}")
    print(f"  F1-score (Macro): {f1_macro:.4f}")
    print(f"  F1-score (Micro): {f1_micro:.4f}")

# Calculate and print the average metrics across all folds
print("\n--- Average Metrics for KNN (k=3) with 5-Fold Cross-Validation ---")
print(f"Average Accuracy: {np.mean(accuracy_scores_knn_3_cv):.4f}")
print(f"Average Precision (Macro): {np.mean(precision_macro_scores_knn_3_cv):.4f}")
print(f"Average Precision (Micro): {np.mean(precision_micro_scores_knn_3_cv):.4f}")
print(f"Average Recall (Macro): {np.mean(recall_macro_scores_knn_3_cv):.4f}")
print(f"Average Recall (Micro): {np.mean(recall_micro_scores_knn_3_cv):.4f}")
print(f"Average F1-score (Macro): {np.mean(f1_macro_scores_knn_3_cv):.4f}")
print(f"Average F1-score (Micro): {np.mean(f1_micro_scores_knn_3_cv):.4f}")

print("\n--- K-Nearest Neighbors (k=5) with 5-Fold Cross-Validation ---")

# Loop through each fold for k=5
for fold, (train_index, val_index) in enumerate(skf.split(X_train_scaled, y_train)):
    X_sub_train, X_val = X_train_scaled[train_index], X_train_scaled[val_index]
    y_sub_train, y_val = y_train.iloc[train_index], y_train.iloc[val_index]

    # Initialize and train KNN (k=5)
    knn_5_cv = KNeighborsClassifier(n_neighbors=5)
    knn_5_cv.fit(X_sub_train, y_sub_train)
    y_pred_knn_5_cv = knn_5_cv.predict(X_val)

    # Calculate metrics for the current fold
    accuracy = accuracy_score(y_val, y_pred_knn_5_cv)
    precision_macro = precision_score(y_val, y_pred_knn_5_cv, average='macro', zero_division=0)
    precision_micro = precision_score(y_val, y_pred_knn_5_cv, average='micro', zero_division=0)
    recall_macro = recall_score(y_val, y_pred_knn_5_cv, average='macro', zero_division=0)
    recall_micro = recall_score(y_val, y_pred_knn_5_cv, average='micro', zero_division=0)
    f1_macro = f1_score(y_val, y_pred_knn_5_cv, average='macro', zero_division=0)
    f1_micro = f1_score(y_val, y_pred_knn_5_cv, average='micro', zero_division=0)

    # Store metrics
    accuracy_scores_knn_5_cv.append(accuracy)
    precision_macro_scores_knn_5_cv.append(precision_macro)
    precision_micro_scores_knn_5_cv.append(precision_micro)
    recall_macro_scores_knn_5_cv.append(recall_macro)
    recall_micro_scores_knn_5_cv.append(recall_micro)
    f1_macro_scores_knn_5_cv.append(f1_macro)
    f1_micro_scores_knn_5_cv.append(f1_micro)

    print(f"\nFold {fold+1}:")
    print(f"  Accuracy: {accuracy:.4f}")
    print(f"  Precision (Macro): {precision_macro:.4f}")
    print(f"  Precision (Micro): {precision_micro:.4f}")
    print(f"  Recall (Macro): {recall_macro:.4f}")
    print(f"  Recall (Micro): {recall_micro:.4f}")
    print(f"  F1-score (Macro): {f1_macro:.4f}")
    print(f"  F1-score (Micro): {f1_micro:.4f}")

# Calculate and print the average metrics across all folds
print("\n--- Average Metrics for KNN (k=5) with 5-Fold Cross-Validation ---")
print(f"Average Accuracy: {np.mean(accuracy_scores_knn_5_cv):.4f}")
print(f"Average Precision (Macro): {np.mean(precision_macro_scores_knn_5_cv):.4f}")
print(f"Average Precision (Micro): {np.mean(precision_micro_scores_knn_5_cv):.4f}")
print(f"Average Recall (Macro): {np.mean(recall_macro_scores_knn_5_cv):.4f}")
print(f"Average Recall (Micro): {np.mean(recall_micro_scores_knn_5_cv):.4f}")
print(f"Average F1-score (Macro): {np.mean(f1_macro_scores_knn_5_cv):.4f}")
print(f"Average F1-score (Micro): {np.mean(f1_micro_scores_knn_5_cv):.4f}")

--- K-Nearest Neighbors (k=3) with 5-Fold Cross-Validation ---

Fold 1:
  Accuracy: 0.9048
  Precision (Macro): 0.9333
  Precision (Micro): 0.9048
  Recall (Macro): 0.9048
  Recall (Micro): 0.9048
  F1-score (Macro): 0.9074
  F1-score (Micro): 0.9048

Fold 2:
  Accuracy: 0.9524
  Precision (Macro): 0.9583
  Precision (Micro): 0.9524
  Recall (Macro): 0.9583
  Recall (Micro): 0.9524
  F1-score (Macro): 0.9556
  F1-score (Micro): 0.9524

Fold 3:
  Accuracy: 0.8571
  Precision (Macro): 0.8750
  Precision (Micro): 0.8571
  Recall (Macro): 0.8552
  Recall (Micro): 0.8571
  F1-score (Macro): 0.8614
  F1-score (Micro): 0.8571

Fold 4:
  Accuracy: 0.9524
  Precision (Macro): 0.9583
  Precision (Micro): 0.9524
  Recall (Macro): 0.9583
  Recall (Micro): 0.9524
  F1-score (Macro): 0.9556
  F1-score (Micro): 0.9524

Fold 5:
  Accuracy: 0.9524
  Precision (Macro): 0.9583
  Precision (Micro): 0.9524
  Recall (Macro): 0.9524
  Recall (Micro): 0.9524
  F1-score (Macro): 0.9521
  F1-score (Micro): 0.95

### b)Compare os resultados da validação cruzada com a divisão simples treino/teste. O que você observa?

**Comparação dos Resultados:**

Na Tarefa 1.1a, tanto o KNN com k=3 quanto o KNN com k=5 apresentaram desempenho perfeito no conjunto de teste, com 100% de Acurácia, Precisão, Recall e F1-score (macro e micro). As matrizes de confusão também mostraram zero erros, indicando que todos os 45 exemplos do conjunto de teste foram classificados corretamente.

Em contraste, os resultados da validação cruzada de 5-folds na Tarefa 1.2a revelaram métricas médias ligeiramente inferiores a 100% para ambos os modelos. Por exemplo, a Acurácia média para k=3 foi de aproximadamente 0.9238 e para k=5 também foi de aproximadamente 0.9238. Além disso, a validação cruzada mostrou variações nas métricas entre os diferentes folds, como visto nos outputs para cada fold.

**Observações:**

A divisão simples treino/teste, embora útil para uma avaliação inicial, pode ser suscetível a overfitting, levando a uma superestimação do desempenho (como o 100% observado). Isso pode ocorrer especialmente em datasets menores ou quando a amostragem do conjunto de teste não é representativa da variabilidade geral dos dados.

A validação cruzada, por outro lado, fornece uma estimativa muito mais robusta e realista do desempenho do modelo em dados não vistos. Ao treinar e testar o modelo múltiplas vezes em diferentes subconjuntos dos dados, ela mitiga o risco de superestimação do desempenho devido a uma divisão de dados particularmente favorável. As métricas médias da validação cruzada, que são um pouco mais baixas, mas consistentes e com uma visão da variabilidade (pelo desvio padrão), oferecem uma indicação mais confiável de como o modelo se comportaria em cenários do mundo real.

### c) Calcule o desvio padrão da acurácia para cada modelo. O que isso indica sobre a estabilidade do modelo?

In [13]:
std_accuracy_knn_3_cv = np.std(accuracy_scores_knn_3_cv)
std_accuracy_knn_5_cv = np.std(accuracy_scores_knn_5_cv)

print(f"Standard Deviation of Accuracy for KNN (k=3): {std_accuracy_knn_3_cv:.4f}")
print(f"Standard Deviation of Accuracy for KNN (k=5): {std_accuracy_knn_5_cv:.4f}")

Standard Deviation of Accuracy for KNN (k=3): 0.0381
Standard Deviation of Accuracy for KNN (k=5): 0.0381


**Desvio Padrão da Acurácia:**

*   **KNN (k=3):** O desvio padrão da acurácia foi de aproximadamente `0.0381`.
*   **KNN (k=5):** O desvio padrão da acurácia foi de aproximadamente `0.0381`.

**Interpretação da Estabilidade do Modelo:**

O desvio padrão da acurácia indica a variabilidade do desempenho do modelo em diferentes folds da validação cruzada. Um desvio padrão menor sugere que o modelo é mais estável e que sua performance é consistente, independentemente da partição dos dados de treino e teste.

Neste caso, ambos os modelos, KNN com k=3 e KNN com k=5, apresentaram um desvio padrão de acurácia idêntico e relativamente baixo (`0.0381`). Isso indica que ambos os modelos são **igualmente estáveis** e demonstram uma performance consistente através das diferentes divisões do conjunto de dados de treino durante a validação cruzada. Eles não mostram uma grande flutuação em sua capacidade de prever corretamente a classe das amostras, o que é um bom indicativo de robustez.

**Conclusão:** Ambos os modelos KNN, com k=3 e k=5, não apenas obtiveram médias de acurácia muito similares (aproximadamente 0.9238), mas também exibem a mesma estabilidade em seus desempenhos de classificação de acordo com o desvio padrão da acurácia. Isso reforça a ideia de que, para este dataset e esta configuração, a escolha entre k=3 e k=5 não resultou em diferenças significativas na estabilidade do modelo.

## Tarefa 1.3: Curvas ROC e AUC

### a) Para o problema binário (Setosa vs. Não-Setosa), plote as curvas ROC para os três modelos.

### b) Calcule a área sob a curva (AUC) para cada modelo.

### c) Interprete os resultados: qual modelo tem melhor capacidade de discriminação?

# Parte 2: REGRESSÃO

---

Dataset: *Boston Housing (ou California Housing)*

Descrição: Predição de preços de imóveis baseado em características socioeconômicas e geográficas.

## Tarefa 2.1: Múltiplos Modelos de Regressão

Implemente o modelo. K-Nearest Neighbors Regressor (k=3 e k=7)

### a) Divida o dataset em 70% treino e 30% teste. Para cada modelo, calcule:

- Erro Médio Absoluto (MAE)
- Erro Quadrático Médio (MSE)
- Raiz do Erro Quadrático Médio (RMSE)
- Coeficiente de Determinação (R²)
- Erro Percentual Absoluto Médio (MAPE)

### b) Crie um gráfico de dispersão comparando valores preditos vs. valores reais para cada modelo.

### c) Qual modelo apresentou melhor desempenho? Por quê?

## Tarefa 2.2: Análise de Resíduos

### a) Para os três melhores modelos, plote o gráfico de resíduos (erro vs. valores preditos).

### b) Há algum padrão nos resíduos? O que isso indica sobre o modelo?

### c) Calcule e plote o histograma dos resíduos. Eles seguem uma distribuição aproximadamente normal?

# Parte 3: CLUSTERIZAÇÃO


---


Dataset: *Wine Dataset*

Descrição: Análise química de vinhos de três cultivares diferentes na Itália.

## Tarefa 3.1: K-Means

### a) Aplique o algoritmo K-Means com k variando de 2 a 10.

In [ ]:
wine = load_wine()
X = wine.data

## Padronização
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


## K-Means variando k de 2 a 10
resultados = []

for k in range(2, 11):
    kmeans = KMeans(n_clusters=k, random_state=42, n_init=10)
    kmeans.fit(X_scaled)

    inertia = kmeans.inertia_  # SSE (erro interno)
    sil = silhouette_score(X_scaled, kmeans.labels_)

    resultados.append([k, inertia, sil])

## mostrar resultados em tabela
df_result = pd.DataFrame(resultados, columns=["k", "Inertia (SSE)", "Silhouette"])
print(df_result)

### b) Para cada valor de k, calcule:

- Inércia (within-cluster sum of squares)
- Silhouette Score
- Davies-Bouldin Index
- Calinski-Harabasz Index

### c) Plote o Método do Cotovelo (Elbow Method) usando a inércia.

### d) Determine o número ótimo de clusters considerando múltiplas métricas.

## Tarefa 3.2: Comparação de Algoritmos
Implemente e compare:

1. K-Means (com k ótimo da Tarefa 3.1)

2. DBSCAN (teste diferentes valores de eps e min_samples)

3. Hierarchical Clustering (Agglomerative)

### a) Calcule o Silhouette Score para cada algoritmo.

### b) Como os labels verdadeiros (cultivares) se comparam com os clusters encontrados?

## Tarefa 3.3: Análise de Perfis dos Clusters

### a) Para a melhor solução de clusterização, calcule:

- Média de cada feature por cluster
- Desvio padrão de cada feature por cluster

### b) Crie um gráfico de radar (spider chart) mostrando o perfil médio de cada cluster.

### c) Descreva as características de cada cluster em termos das features originais.